{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Before starting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Copy the protein fasta file of the organism in the directory \"/data/annotations/fasta/\"\n",
    "- Be sure you have the trained deep learning models to do the compartment prediction of the proteins in the folder \"/bin/compartmentPredictions/deepModels\". It can be downloaded from \"http://doi.org/10.5281/zenodo.4436488\"\n",
    "- Install the eggNog annotation software: https://github.com/eggnogdb/eggnog-mapper\n",
    "- Install the python libraries:Tensorflow, keras, biopython, pandas, numpy, cobrapy, Framed (https://github.com/cdanielmachado/framed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functional annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "download eggnog-mapper from https://github.com/eggnogdb/eggnog-mapper. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "import subprocess"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "python2_command= \"bin/eggnog-mapper-master/emapper.py -i \\\"data/annotations/fasta/example.fasta\\\" --database euk --output \\\"data/annotations/functional/example.fasta\\\" -m diamond --override --cpu 32\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "process = subprocess.Popen(python2_command.split(), stdout=subprocess.PIPE)\n",
    "output, error = process.communicate() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "b'/usr/local/bin/diamond /mnt/home/scsandra/proj/reconstructionPipeline/bin/eggnog-mapper-master\\n'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "error"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compartment prediction of the proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#query_name</th>\n",
       "      <th>seed_eggNOG_ortholog</th>\n",
       "      <th>seed_ortholog_evalue</th>\n",
       "      <th>seed_ortholog_score</th>\n",
       "      <th>best_tax_level</th>\n",
       "      <th>Preferred_name</th>\n",
       "      <th>GOs</th>\n",
       "      <th>EC</th>\n",
       "      <th>KEGG_ko</th>\n",
       "      <th>KEGG_Pathway</th>\n",
       "      <th>...</th>\n",
       "      <th>KEGG_rclass</th>\n",
       "      <th>BRITE</th>\n",
       "      <th>KEGG_TC</th>\n",
       "      <th>CAZy</th>\n",
       "      <th>BiGG_Reaction</th>\n",
       "      <th>taxonomic scope</th>\n",
       "      <th>eggNOG OGs</th>\n",
       "      <th>best eggNOG OG</th>\n",
       "      <th>COG Functional cat.</th>\n",
       "      <th>eggNOG free text desc.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NP_001018030.1</td>\n",
       "      <td>4932.YCL064C</td>\n",
       "      <td>2.900000e-204</td>\n",
       "      <td>717.6</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>CHA1</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0003941,GO:0004794,GO...</td>\n",
       "      <td>4.3.1.17,4.3.1.19</td>\n",
       "      <td>ko:K17989</td>\n",
       "      <td>ko00260,ko00270,ko00290,ko01100,ko01110,ko0113...</td>\n",
       "      <td>...</td>\n",
       "      <td>RC00331,RC00418,RC02600</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>38BCJ@33154,3NZW4@4751,3QNQX@4890,3RR5U@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>E</td>\n",
       "      <td>to Saccharomyces cerevisiae CHA1 (YCL064C)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NP_001335748.1</td>\n",
       "      <td>4932.YCR001W</td>\n",
       "      <td>1.600000e-54</td>\n",
       "      <td>218.4</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...</td>\n",
       "      <td>4.1.3.1</td>\n",
       "      <td>ko:K01637</td>\n",
       "      <td>ko00630,ko01100,ko01110,ko01120,ko01200,map006...</td>\n",
       "      <td>...</td>\n",
       "      <td>RC00311,RC00313</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>COG0038@1,KOG0475@2759</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>P</td>\n",
       "      <td>voltage-gated chloride channel activity</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NP_009305.1</td>\n",
       "      <td>4932.Q0045</td>\n",
       "      <td>7.600000e-302</td>\n",
       "      <td>1042.3</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>cox1</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0004129,GO:0005215,GO...</td>\n",
       "      <td>1.9.3.1</td>\n",
       "      <td>ko:K02256</td>\n",
       "      <td>ko00190,ko01100,ko04260,ko04714,ko04932,ko0501...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000,ko03029</td>\n",
       "      <td>3.D.4.11,3.D.4.7,3.D.4.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>38JQ7@33154,3NYGG@4751,3QSKE@4890,3RU36@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>C</td>\n",
       "      <td>electron transport coupled proton transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NP_009319.1</td>\n",
       "      <td>4932.Q0130</td>\n",
       "      <td>2.600000e-30</td>\n",
       "      <td>137.5</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>atp9</td>\n",
       "      <td>GO:0000276,GO:0003674,GO:0003824,GO:0005215,GO...</td>\n",
       "      <td>1.6.5.3</td>\n",
       "      <td>ko:K02128,ko:K03880</td>\n",
       "      <td>ko00190,ko01100,ko04714,ko04723,ko05010,ko0501...</td>\n",
       "      <td>...</td>\n",
       "      <td>RC00061</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000,ko03029</td>\n",
       "      <td>3.A.2.1,3.D.1.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>3A3YG@33154,3P7BC@4751,3QYEK@4890,3RV97@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>C</td>\n",
       "      <td>ATP hydrolysis coupled proton transport</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NP_009339.1</td>\n",
       "      <td>4932.YAL062W</td>\n",
       "      <td>1.500000e-261</td>\n",
       "      <td>908.3</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>GDH3</td>\n",
       "      <td>GO:0001505,GO:0003674,GO:0003824,GO:0004353,GO...</td>\n",
       "      <td>1.4.1.4</td>\n",
       "      <td>ko:K00262</td>\n",
       "      <td>ko00220,ko00250,ko00910,ko01100,map00220,map00...</td>\n",
       "      <td>...</td>\n",
       "      <td>RC00006,RC02799</td>\n",
       "      <td>ko00000,ko00001,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>38GEK@33154,3NVMC@4751,3QKVX@4890,3RRZ0@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>E</td>\n",
       "      <td>Belongs to the Glu Leu Phe Val dehydrogenases ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5532</th>\n",
       "      <td>NP_116708.1</td>\n",
       "      <td>4932.YFR050C</td>\n",
       "      <td>7.400000e-149</td>\n",
       "      <td>533.1</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>PRE4</td>\n",
       "      <td>GO:0000502,GO:0005575,GO:0005622,GO:0005623,GO...</td>\n",
       "      <td>3.4.25.1</td>\n",
       "      <td>ko:K02736</td>\n",
       "      <td>ko03050,map03050</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000,ko01002,ko03051</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>38G4U@33154,3NUK4@4751,3QJEW@4890,3RSRP@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>O</td>\n",
       "      <td>Belongs to the peptidase T1B family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5535</th>\n",
       "      <td>NP_116711.3</td>\n",
       "      <td>4932.YFR053C</td>\n",
       "      <td>7.900000e-282</td>\n",
       "      <td>975.7</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>hxk1</td>\n",
       "      <td>GO:0000166,GO:0001302,GO:0001678,GO:0003674,GO...</td>\n",
       "      <td>2.7.1.1</td>\n",
       "      <td>ko:K00844</td>\n",
       "      <td>ko00010,ko00051,ko00052,ko00500,ko00520,ko0052...</td>\n",
       "      <td>...</td>\n",
       "      <td>RC00002,RC00017</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000,ko04131</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>iMM904.YFR053C,iND750.YFR053C</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>38DFC@33154,3NW32@4751,3QNBF@4890,3RR3P@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>G</td>\n",
       "      <td>Belongs to the hexokinase family</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5536</th>\n",
       "      <td>NP_116714.1</td>\n",
       "      <td>4932.YFR055W</td>\n",
       "      <td>5.500000e-197</td>\n",
       "      <td>693.3</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0000096,GO:0000098,GO:0000724,GO:0000725,GO...</td>\n",
       "      <td>4.4.1.8</td>\n",
       "      <td>ko:K01760</td>\n",
       "      <td>ko00270,ko00450,ko01100,ko01110,ko01230,map002...</td>\n",
       "      <td>...</td>\n",
       "      <td>RC00056,RC00069,RC00382,RC00488,RC00710,RC0124...</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>39WRI@33154,3PCP5@4751,3R1A3@4890,3RW6C@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>E</td>\n",
       "      <td>Cys/Met metabolism PLP-dependent enzyme</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5544</th>\n",
       "      <td>NP_602317.3</td>\n",
       "      <td>4932.YFR047C</td>\n",
       "      <td>2.900000e-170</td>\n",
       "      <td>604.4</td>\n",
       "      <td>Saccharomycetaceae</td>\n",
       "      <td>BNA6</td>\n",
       "      <td>GO:0003674,GO:0003824,GO:0004514,GO:0005575,GO...</td>\n",
       "      <td>2.4.2.19</td>\n",
       "      <td>ko:K00767</td>\n",
       "      <td>ko00760,ko01100,map00760,map01100</td>\n",
       "      <td>...</td>\n",
       "      <td>RC02877</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Fungi</td>\n",
       "      <td>38D18@33154,3NUQX@4751,3QMBE@4890,3RS6P@4891,3...</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>H</td>\n",
       "      <td>Involved in the catabolism of quinolinic acid ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5563</th>\n",
       "      <td>NP_878063.3</td>\n",
       "      <td>4932.YDR182W-A</td>\n",
       "      <td>6.000000e-31</td>\n",
       "      <td>139.4</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>RCH1</td>\n",
       "      <td>GO:0003674,GO:0005215,GO:0005319,GO:0005575,GO...</td>\n",
       "      <td>3.4.25.1</td>\n",
       "      <td>ko:K02731,ko:K14347</td>\n",
       "      <td>ko03050,map03050</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ko00000,ko00001,ko00002,ko01000,ko01002,ko0200...</td>\n",
       "      <td>2.A.93.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>COG0385@1,KOG4821@2759</td>\n",
       "      <td>NA|NA|NA</td>\n",
       "      <td>J</td>\n",
       "      <td>symporter activity</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1391 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #query_name seed_eggNOG_ortholog  seed_ortholog_evalue  \\\n",
       "0     NP_001018030.1         4932.YCL064C         2.900000e-204   \n",
       "11    NP_001335748.1         4932.YCR001W          1.600000e-54   \n",
       "14       NP_009305.1           4932.Q0045         7.600000e-302   \n",
       "27       NP_009319.1           4932.Q0130          2.600000e-30   \n",
       "37       NP_009339.1         4932.YAL062W         1.500000e-261   \n",
       "...              ...                  ...                   ...   \n",
       "5532     NP_116708.1         4932.YFR050C         7.400000e-149   \n",
       "5535     NP_116711.3         4932.YFR053C         7.900000e-282   \n",
       "5536     NP_116714.1         4932.YFR055W         5.500000e-197   \n",
       "5544     NP_602317.3         4932.YFR047C         2.900000e-170   \n",
       "5563     NP_878063.3       4932.YDR182W-A          6.000000e-31   \n",
       "\n",
       "      seed_ortholog_score      best_tax_level Preferred_name  \\\n",
       "0                   717.6  Saccharomycetaceae           CHA1   \n",
       "11                  218.4           Eukaryota            NaN   \n",
       "14                 1042.3  Saccharomycetaceae           cox1   \n",
       "27                  137.5  Saccharomycetaceae           atp9   \n",
       "37                  908.3  Saccharomycetaceae           GDH3   \n",
       "...                   ...                 ...            ...   \n",
       "5532                533.1  Saccharomycetaceae           PRE4   \n",
       "5535                975.7  Saccharomycetaceae           hxk1   \n",
       "5536                693.3  Saccharomycetaceae            NaN   \n",
       "5544                604.4  Saccharomycetaceae           BNA6   \n",
       "5563                139.4           Eukaryota           RCH1   \n",
       "\n",
       "                                                    GOs                 EC  \\\n",
       "0     GO:0003674,GO:0003824,GO:0003941,GO:0004794,GO...  4.3.1.17,4.3.1.19   \n",
       "11    GO:0005575,GO:0005622,GO:0005623,GO:0005737,GO...            4.1.3.1   \n",
       "14    GO:0003674,GO:0003824,GO:0004129,GO:0005215,GO...            1.9.3.1   \n",
       "27    GO:0000276,GO:0003674,GO:0003824,GO:0005215,GO...            1.6.5.3   \n",
       "37    GO:0001505,GO:0003674,GO:0003824,GO:0004353,GO...            1.4.1.4   \n",
       "...                                                 ...                ...   \n",
       "5532  GO:0000502,GO:0005575,GO:0005622,GO:0005623,GO...           3.4.25.1   \n",
       "5535  GO:0000166,GO:0001302,GO:0001678,GO:0003674,GO...            2.7.1.1   \n",
       "5536  GO:0000096,GO:0000098,GO:0000724,GO:0000725,GO...            4.4.1.8   \n",
       "5544  GO:0003674,GO:0003824,GO:0004514,GO:0005575,GO...           2.4.2.19   \n",
       "5563  GO:0003674,GO:0005215,GO:0005319,GO:0005575,GO...           3.4.25.1   \n",
       "\n",
       "                  KEGG_ko                                       KEGG_Pathway  \\\n",
       "0               ko:K17989  ko00260,ko00270,ko00290,ko01100,ko01110,ko0113...   \n",
       "11              ko:K01637  ko00630,ko01100,ko01110,ko01120,ko01200,map006...   \n",
       "14              ko:K02256  ko00190,ko01100,ko04260,ko04714,ko04932,ko0501...   \n",
       "27    ko:K02128,ko:K03880  ko00190,ko01100,ko04714,ko04723,ko05010,ko0501...   \n",
       "37              ko:K00262  ko00220,ko00250,ko00910,ko01100,map00220,map00...   \n",
       "...                   ...                                                ...   \n",
       "5532            ko:K02736                                   ko03050,map03050   \n",
       "5535            ko:K00844  ko00010,ko00051,ko00052,ko00500,ko00520,ko0052...   \n",
       "5536            ko:K01760  ko00270,ko00450,ko01100,ko01110,ko01230,map002...   \n",
       "5544            ko:K00767                  ko00760,ko01100,map00760,map01100   \n",
       "5563  ko:K02731,ko:K14347                                   ko03050,map03050   \n",
       "\n",
       "      ...                                        KEGG_rclass  \\\n",
       "0     ...                            RC00331,RC00418,RC02600   \n",
       "11    ...                                    RC00311,RC00313   \n",
       "14    ...                                                NaN   \n",
       "27    ...                                            RC00061   \n",
       "37    ...                                    RC00006,RC02799   \n",
       "...   ...                                                ...   \n",
       "5532  ...                                                NaN   \n",
       "5535  ...                                    RC00002,RC00017   \n",
       "5536  ...  RC00056,RC00069,RC00382,RC00488,RC00710,RC0124...   \n",
       "5544  ...                                            RC02877   \n",
       "5563  ...                                                NaN   \n",
       "\n",
       "                                                  BRITE  \\\n",
       "0                       ko00000,ko00001,ko00002,ko01000   \n",
       "11                      ko00000,ko00001,ko00002,ko01000   \n",
       "14              ko00000,ko00001,ko00002,ko01000,ko03029   \n",
       "27              ko00000,ko00001,ko00002,ko01000,ko03029   \n",
       "37                              ko00000,ko00001,ko01000   \n",
       "...                                                 ...   \n",
       "5532    ko00000,ko00001,ko00002,ko01000,ko01002,ko03051   \n",
       "5535            ko00000,ko00001,ko00002,ko01000,ko04131   \n",
       "5536                    ko00000,ko00001,ko00002,ko01000   \n",
       "5544                    ko00000,ko00001,ko00002,ko01000   \n",
       "5563  ko00000,ko00001,ko00002,ko01000,ko01002,ko0200...   \n",
       "\n",
       "                       KEGG_TC CAZy                  BiGG_Reaction  \\\n",
       "0                          NaN  NaN                            NaN   \n",
       "11                         NaN  NaN                            NaN   \n",
       "14    3.D.4.11,3.D.4.7,3.D.4.8  NaN                            NaN   \n",
       "27             3.A.2.1,3.D.1.6  NaN                            NaN   \n",
       "37                         NaN  NaN                            NaN   \n",
       "...                        ...  ...                            ...   \n",
       "5532                       NaN  NaN                            NaN   \n",
       "5535                       NaN  NaN  iMM904.YFR053C,iND750.YFR053C   \n",
       "5536                       NaN  NaN                            NaN   \n",
       "5544                       NaN  NaN                            NaN   \n",
       "5563                  2.A.93.1  NaN                            NaN   \n",
       "\n",
       "     taxonomic scope                                         eggNOG OGs  \\\n",
       "0              Fungi  38BCJ@33154,3NZW4@4751,3QNQX@4890,3RR5U@4891,3...   \n",
       "11         Eukaryota                             COG0038@1,KOG0475@2759   \n",
       "14             Fungi  38JQ7@33154,3NYGG@4751,3QSKE@4890,3RU36@4891,3...   \n",
       "27             Fungi  3A3YG@33154,3P7BC@4751,3QYEK@4890,3RV97@4891,3...   \n",
       "37             Fungi  38GEK@33154,3NVMC@4751,3QKVX@4890,3RRZ0@4891,3...   \n",
       "...              ...                                                ...   \n",
       "5532           Fungi  38G4U@33154,3NUK4@4751,3QJEW@4890,3RSRP@4891,3...   \n",
       "5535           Fungi  38DFC@33154,3NW32@4751,3QNBF@4890,3RR3P@4891,3...   \n",
       "5536           Fungi  39WRI@33154,3PCP5@4751,3R1A3@4890,3RW6C@4891,3...   \n",
       "5544           Fungi  38D18@33154,3NUQX@4751,3QMBE@4890,3RS6P@4891,3...   \n",
       "5563       Eukaryota                             COG0385@1,KOG4821@2759   \n",
       "\n",
       "     best eggNOG OG COG Functional cat.  \\\n",
       "0          NA|NA|NA                   E   \n",
       "11         NA|NA|NA                   P   \n",
       "14         NA|NA|NA                   C   \n",
       "27         NA|NA|NA                   C   \n",
       "37         NA|NA|NA                   E   \n",
       "...             ...                 ...   \n",
       "5532       NA|NA|NA                   O   \n",
       "5535       NA|NA|NA                   G   \n",
       "5536       NA|NA|NA                   E   \n",
       "5544       NA|NA|NA                   H   \n",
       "5563       NA|NA|NA                   J   \n",
       "\n",
       "                                 eggNOG free text desc.  \n",
       "0            to Saccharomyces cerevisiae CHA1 (YCL064C)  \n",
       "11              voltage-gated chloride channel activity  \n",
       "14          electron transport coupled proton transport  \n",
       "27              ATP hydrolysis coupled proton transport  \n",
       "37    Belongs to the Glu Leu Phe Val dehydrogenases ...  \n",
       "...                                                 ...  \n",
       "5532                Belongs to the peptidase T1B family  \n",
       "5535                   Belongs to the hexokinase family  \n",
       "5536            Cys/Met metabolism PLP-dependent enzyme  \n",
       "5544  Involved in the catabolism of quinolinic acid ...  \n",
       "5563                                 symporter activity  \n",
       "\n",
       "[1391 rows x 22 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "functional_annotation= pd.read_csv(\"data/annotations/functional/Saccharomyces_cerevisiae_S288C_GCF_000146045.2_NEWprotein.faa.emapper.annotations\",skiprows=3,sep=\"\\t\")\n",
    "functional_annotation = functional_annotation[functional_annotation['EC'].notnull()]\n",
    "functional_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "import tempfile\n",
    "import shutil\n",
    "\n",
    "dirpath = tempfile.mkdtemp()\n",
    "# ... do stuff with dirpath\n",
    "#shutil.rmtree(dirpath)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "selected=[]\n",
    "for record in SeqIO.parse(\"data/annotations/fasta/Saccharomyces_cerevisiae_S288C_GCF_000146045.2_NEWprotein.faa\", \"fasta\"):\n",
    "    sel = functional_annotation[functional_annotation[\"#query_name\"]==record.id]\n",
    "    if len(sel) >0:\n",
    "        SeqIO.write(record, dirpath+\"/\"+record.id+\".fasta\", \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/tmp/tmp_fk_dfss\n"
     ]
    }
   ],
   "source": [
    "print(dirpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, '3rd_party/psipred')\n",
    "import subprocess\n",
    "\n",
    "for key, row in functional_annotation.iterrows():\n",
    "    process = subprocess.Popen(['3rd_party/psipred/runSec.sh', dirpath+\"/\"+row[\"#query_name\"]],\n",
    "                         stdout=subprocess.PIPE, \n",
    "                         stderr=subprocess.PIPE)\n",
    "    stdout, stderr = process.communicate()\n",
    "    stdout, stderr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "dirpath =\"/tmp/tmp_fk_dfss\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1, 500, 3)\n",
      "(360, 7)\n",
      "(1, 500, 7)\n",
      "(1, 500, 21)\n",
      "(1, 500, 3)\n",      
      "predicting\n",
      "(1391, 4)\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.insert(0, 'bin/compartmentPrediction/')\n",
    "import predict\n",
    "predict.predict(dirpath+\"/\", \"Saccharomyces_cerevisiae_S288C_GCF_000146045.2_NEWprotein\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shutil.rmtree(dirpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Running CarveMe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, 'bin/')\n",
    "import CreateModelEggNogPool "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import logging\n",
    "serf_logger = logging.getLogger('cobra')\n",
    "serf_logger.setLevel(logging.ERROR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Scores calculated\n",
      "Solution 0\n",
      "Objective: 808.3812231553678\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 0.907626658490266\n",
      "Status: Optimal\n",
      "\n",
      "Solution 1\n",
      "Objective: 808.3812231552508\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 0.9073117116060194\n",
      "Status: Optimal\n",
      "\n",
      "Solution 2\n",
      "Objective: 808.3812231551436\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 0.5939739728870619\n",
      "Status: Optimal\n",
      "\n",
      "Solution 3\n",
      "Objective: 808.0331473085121\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 0.8213937982287794\n",
      "Status: Optimal\n",
      "\n",
      "Solution 4\n",
      "Objective: 808.033147308503\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 0.8213937982288066\n",
      "Status: Optimal\n",
      "\n",
      "Solution 0\n",
      "Objective: 922.5048103424522\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 8.466720230026676\n",
      "Status: Optimal\n",
      "\n",
      "Solution 1\n",
      "Objective: 921.9217989186573\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 7.612555546634004\n",
      "Status: Optimal\n",
      "\n",
      "Solution 2\n",
      "Objective: 921.3685467393694\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 7.718138574643925\n",
      "Status: Optimal\n",
      "\n",
      "Solution 3\n",
      "Objective: 921.3685467393104\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 7.002076850335239\n",
      "Status: Optimal\n",
      "\n",
      "Solution 4\n",
      "Objective: 921.1638039657746\n",
      "Status: Unknown\n",
      "\n",
      "Objective: 8.194226464511308\n",
      "Status: Optimal\n",
      "\n",
      "UF03012_E e exchange\n",
      "UF03013_E e exchange\n",
      "UF03147_TCE e export\n",
      "UF03072_E e exchange\n",
      "UF02958_E e exchange\n",
      "UF03014_E e exchange\n",
      "UF03012_E e exchange\n",
      "UF03013_E e exchange\n",
      "UF03147_TCE e export\n",
      "UF03072_E e exchange\n",
      "UF02958_E e exchange\n",
      "UF03014_E e exchange\n",
      "Done 808.2419928165555 921.6655013411128\n"
     ]
    }
   ],
   "source": [
    "CreateModelEggNogPool.carveme_pipeline(\"Scer_v21\", \"Saccharomyces_cerevisiae_S288C_GCF_000146045.2\",\"data/annotations/functional/Saccharomyces_cerevisiae_S288C_GCF_000146045.2_NEWprotein.faa.emapper.annotations\",\"data/annotations/fasta/Saccharomyces_cerevisiae_S288C_GCF_000146045.2_NEWprotein.faa\",\"data/annotations/localization/Saccharomyces_cerevisiae_S288C_GCF_000146045.2_NEWprotein.loc_pred\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting model contains two numbers corresponding to the objective result of the two MILP problems (models with open and closed bounds)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember to create few models and selecte the one with bigger objective results. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analyzing the resulting model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "import cobra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = cobra.io.read_sbml_model(\"results/Scer_v21-808.2419928165555-921.6655013411128.sbml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Fixing the BIOMASS reaction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra import Model, Reaction, Metabolite\n",
    "reaction = Reaction('BIOMASS')\n",
    "reaction.name = model.reactions.BIOMASS.name\n",
    "reaction.add_metabolites({\n",
    "    model.metabolites.get_by_id(\"13BDglcn_1g_c\"): -0.253,\n",
    "    model.metabolites.ash_1g_c: -0.048,\n",
    "    model.metabolites.atp_c: -60.0,\n",
    "    model.metabolites.chitin_1g_c: -0.01,\n",
    "    model.metabolites.dna_1g_c: -0.005,\n",
    "    model.metabolites.glycogen_1g_c: -0.1,\n",
    "    model.metabolites.h2o_c: -60.0,\n",
    "    model.metabolites.mannan_1g_c: -0.07,\n",
    "    model.metabolites.phospholipid_1g_c: -0.04,\n",
    "    model.metabolites.protein_1g_c: -0.4,\n",
    "    model.metabolites.rna_1g_c: -0.058,\n",
    "    model.metabolites.sterol_1g_r: -0.001,\n",
    "    model.metabolites.trehalose_1g_c: -0.015,\n",
    "    model.metabolites.ficytc_c: -0.0001,\n",
    "    model.metabolites.focytc_c: -0.0001,\n",
    "    model.metabolites.retn_c: -0.0001,\n",
    "    model.metabolites.thmtp_c: -0.0001,\n",
    "    model.metabolites.ribflv_c: -0.0001,\n",
    "    model.metabolites.btn_c: -0.0001,\n",
    "    model.metabolites.q6_m: -0.0001,\n",
    "    model.metabolites.coa_c: -0.0001,\n",
    "    model.metabolites.thf_c: -0.0001,\n",
    "    model.metabolites.adp_c: 60.0,\n",
    "    model.metabolites.biomass_1g_c: 1.0,\n",
    "    model.metabolites.h_c: 60.0,\n",
    "    model.metabolites.pi_c: 60.0\n",
    "})\n",
    "\n",
    "model.reactions.BIOMASS.remove_from_model()\n",
    "model.add_reaction(reaction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.949</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>UF00005_C</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF00014_C</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF00015_C</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.091292e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF00016_C</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF00017_C</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.091292e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF01018_M</th>\n",
       "      <td>0.102631</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF02331_M</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-2.775558e-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF01998_M</th>\n",
       "      <td>0.102631</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UF00028_M</th>\n",
       "      <td>0.102631</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BIOMASS</th>\n",
       "      <td>0.949424</td>\n",
       "      <td>-6.661338e-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1912 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.949 at 0x2b0b58e04cf8>"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.objective= \"BIOMASS\"\n",
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "cobra.io.write_sbml_model(model,\"results/Scer_v21-808.2419928165555-921.6655013411128.sbml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Analyze individual models in the assembly"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def extract_single_model_from_ensemble(position, model):\n",
    "    newModel = model.copy()\n",
    "    reactiond = {}\n",
    "    for reaction in model.reactions:\n",
    "        name = reaction.name.split('_')       \n",
    "        ensemble=[]\n",
    "        for n in name[1:]:\n",
    "            try:\n",
    "                n = int(n)\n",
    "                ensemble.append(n) \n",
    "            except:\n",
    "                None\n",
    "        reactiond[reaction.id]=ensemble       \n",
    "    \n",
    "    for reaction in model.reactions:\n",
    "        presence=reactiond[reaction.id]          \n",
    "        if presence[position]==0:           \n",
    "            newModel.remove_reactions([reaction.id])\n",
    "            \n",
    "    return newModel \n",
    "def get_compartments_rxn(reaction):\n",
    "    comp =[]\n",
    "    for metabolite in reaction.metabolites:\n",
    "        comp.append(metabolite.compartment)\n",
    "    return list(set(comp))\n",
    "\n",
    "def get_compartments(model):\n",
    "    cyt=0\n",
    "    nucleus=0\n",
    "    ret=0\n",
    "    mit=0\n",
    "    per=0\n",
    "    for reaction in model.reactions:\n",
    "        compartments= get_compartments_rxn(reaction)\n",
    "        for c in compartments:\n",
    "            if 'c' in c:\n",
    "                cyt = cyt+1\n",
    "            elif 'r' in c:\n",
    "                ret = ret+1\n",
    "            elif 'x' in c:\n",
    "                per=per+1\n",
    "            elif 'n' in c:\n",
    "                nucleus=nucleus+1\n",
    "            elif 'm' in c:\n",
    "                mit =mit+1\n",
    "    return [cyt, nucleus, ret, mit, per]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model 0\n",
      "Growth: 0.9484703842864195\n",
      "Number of reactions: 1811\n",
      "Number of compounds: 1461\n",
      "Reactions in cytoplasm: 1132\n",
      "Reactions in endoplasmic reticulum: 65\n",
      "Reactions in mitochondria: 234\n",
      "Reactions in peroxisome: 292\n",
      "--------------------------------\n",
      "Model 1\n",
      "Growth: 0.9484703842864297\n",
      "Number of reactions: 1763\n",
      "Number of compounds: 1461\n",
      "Reactions in cytoplasm: 1107\n",
      "Reactions in endoplasmic reticulum: 65\n",
      "Reactions in mitochondria: 236\n",
      "Reactions in peroxisome: 292\n",
      "--------------------------------\n",
      "Model 2\n",
      "Growth: 0.9490855501586049\n",
      "Number of reactions: 1786\n",
      "Number of compounds: 1461\n",
      "Reactions in cytoplasm: 1120\n",
      "Reactions in endoplasmic reticulum: 65\n",
      "Reactions in mitochondria: 239\n",
      "Reactions in peroxisome: 292\n",
      "--------------------------------\n",
      "Model 3\n",
      "Growth: 0.9490855501586178\n",
      "Number of reactions: 1786\n",
      "Number of compounds: 1461\n",
      "Reactions in cytoplasm: 1119\n",
      "Reactions in endoplasmic reticulum: 65\n",
      "Reactions in mitochondria: 238\n",
      "Reactions in peroxisome: 292\n",
      "--------------------------------\n",
      "Model 4\n",
      "Growth: 0.9484703842864144\n",
      "Number of reactions: 1833\n",
      "Number of compounds: 1461\n",
      "Reactions in cytoplasm: 1136\n",
      "Reactions in endoplasmic reticulum: 62\n",
      "Reactions in mitochondria: 233\n",
      "Reactions in peroxisome: 311\n",
      "--------------------------------\n",
      "Model 5\n",
      "Growth: 0.948470384286412\n",
      "Number of reactions: 1804\n",
      "Number of compounds: 1461\n",
      "Reactions in cytoplasm: 1127\n",
      "Reactions in endoplasmic reticulum: 65\n",
      "Reactions in mitochondria: 227\n",
      "Reactions in peroxisome: 292\n",
      "--------------------------------\n",
      "Model 6\n",
      "Growth: 0.9484703842864142\n",
      "Number of reactions: 1747\n",
      "Number of compounds: 1461\n",
      "Reactions in cytoplasm: 1097\n",
      "Reactions in endoplasmic reticulum: 65\n",
      "Reactions in mitochondria: 229\n",
      "Reactions in peroxisome: 292\n",
      "--------------------------------\n"
     ]
    }
   ],
   "source": [
    "for i in range(25):\n",
    "    newModel = extract_single_model_from_ensemble(i, model)\n",
    "    print(\"Model \" +str(i))\n",
    "    print(\"Growth: \" + str(newModel.optimize().objective_value))\n",
    "    print(\"Number of reactions: \" + str(len(newModel.reactions)))\n",
    "    print(\"Number of compounds: \" + str(len(newModel.metabolites)))\n",
    "    [cyt, nucleus, ret, mit, per]=get_compartments(newModel)\n",
    "    print(\"Reactions in cytoplasm: \" + str(cyt))\n",
    "    print(\"Reactions in endoplasmic reticulum: \" + str(ret))\n",
    "    print(\"Reactions in mitochondria: \" + str(mit))\n",
    "    print(\"Reactions in peroxisome: \" + str(per))\n",
    "    print(\"--------------------------------\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model 0\n",
      "Growth: 0.9506473077373335\n",
      "Number of reactions: 1913\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1140\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 321\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 1\n",
      "Growth: 0.9503148490733468\n",
      "Number of reactions: 1909\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1138\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 320\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 2\n",
      "Growth: 0.9503148490733484\n",
      "Number of reactions: 1904\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1134\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 318\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 3\n",
      "Growth: 0.9503148490733471\n",
      "Number of reactions: 1904\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1134\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 319\n",
      "Reactions in peroxisome: 268\n",
      "--------------------------------\n",
      "Model 4\n",
      "Growth: 0.9503148490733428\n",
      "Number of reactions: 1905\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1135\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 319\n",
      "Reactions in peroxisome: 268\n",
      "--------------------------------\n",
      "Model 5\n",
      "Growth: 0.9374212070346898\n",
      "Number of reactions: 1912\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1138\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 320\n",
      "Reactions in peroxisome: 270\n",
      "--------------------------------\n",
      "Model 6\n",
      "Growth: 0.9350166851330837\n",
      "Number of reactions: 1909\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1137\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 319\n",
      "Reactions in peroxisome: 270\n",
      "--------------------------------\n",
      "Model 7\n",
      "Growth: 0.9387948639516787\n",
      "Number of reactions: 1904\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1133\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 317\n",
      "Reactions in peroxisome: 270\n",
      "--------------------------------\n",
      "Model 8\n",
      "Growth: 0.9503148490733455\n",
      "Number of reactions: 1902\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1133\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 316\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 9\n",
      "Growth: 0.9350166851331124\n",
      "Number of reactions: 1900\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1131\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 314\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 10\n",
      "Growth: 0.950647307737297\n",
      "Number of reactions: 1912\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1140\n",
      "Reactions in endoplasmic reticulum: 85\n",
      "Reactions in mitochondria: 320\n",
      "Reactions in peroxisome: 270\n",
      "--------------------------------\n",
      "Model 11\n",
      "Growth: 0.9503148490733793\n",
      "Number of reactions: 1909\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1139\n",
      "Reactions in endoplasmic reticulum: 85\n",
      "Reactions in mitochondria: 320\n",
      "Reactions in peroxisome: 270\n",
      "--------------------------------\n",
      "Model 12\n",
      "Growth: 0.9503148490733426\n",
      "Number of reactions: 1905\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1136\n",
      "Reactions in endoplasmic reticulum: 85\n",
      "Reactions in mitochondria: 319\n",
      "Reactions in peroxisome: 270\n",
      "--------------------------------\n",
      "Model 13\n",
      "Growth: 0.9503148490733511\n",
      "Number of reactions: 1904\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1135\n",
      "Reactions in endoplasmic reticulum: 85\n",
      "Reactions in mitochondria: 319\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 14\n",
      "Growth: 0.9503148490733381\n",
      "Number of reactions: 1904\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1135\n",
      "Reactions in endoplasmic reticulum: 85\n",
      "Reactions in mitochondria: 318\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 15\n",
      "Growth: 0.9374207020013336\n",
      "Number of reactions: 1909\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1136\n",
      "Reactions in endoplasmic reticulum: 85\n",
      "Reactions in mitochondria: 319\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 16\n",
      "Growth: 0.9350166851331134\n",
      "Number of reactions: 1907\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1136\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 318\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 17\n",
      "Growth: 0.9387948639516615\n",
      "Number of reactions: 1903\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1133\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 317\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 18\n",
      "Growth: 0.9503148490733418\n",
      "Number of reactions: 1901\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1133\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 316\n",
      "Reactions in peroxisome: 268\n",
      "--------------------------------\n",
      "Model 19\n",
      "Growth: 0.9350166851330909\n",
      "Number of reactions: 1899\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1131\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 314\n",
      "Reactions in peroxisome: 268\n",
      "--------------------------------\n",
      "Model 20\n",
      "Growth: 0.950647307737305\n",
      "Number of reactions: 1912\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1139\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 321\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 21\n",
      "Growth: 0.9503148490733467\n",
      "Number of reactions: 1908\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1137\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 320\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 22\n",
      "Growth: 0.9503148490733547\n",
      "Number of reactions: 1903\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1133\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 318\n",
      "Reactions in peroxisome: 269\n",
      "--------------------------------\n",
      "Model 23\n",
      "Growth: 0.9503148490733453\n",
      "Number of reactions: 1901\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1133\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 317\n",
      "Reactions in peroxisome: 268\n",
      "--------------------------------\n",
      "Model 24\n",
      "Growth: 0.9503148490733381\n",
      "Number of reactions: 1901\n",
      "Number of compounds: 1493\n",
      "Reactions in cytoplasm: 1133\n",
      "Reactions in endoplasmic reticulum: 86\n",
      "Reactions in mitochondria: 316\n",
      "Reactions in peroxisome: 268\n",
      "--------------------------------\n"
     ]
    }
   ],
   "source": [
    "for i in range(25):\n",
    "    newModel = extract_single_model_from_ensemble(i, model)\n",
    "    print(\"Model \" +str(i))\n",
    "    print(\"Growth: \" + str(newModel.optimize().objective_value))\n",
    "    print(\"Number of reactions: \" + str(len(newModel.reactions)))\n",
    "    print(\"Number of compounds: \" + str(len(newModel.metabolites)))\n",
    "    [cyt, nucleus, ret, mit, per]=get_compartments(newModel)\n",
    "    print(\"Reactions in cytoplasm: \" + str(cyt))\n",
    "    print(\"Reactions in endoplasmic reticulum: \" + str(ret))\n",
    "    print(\"Reactions in mitochondria: \" + str(mit))\n",
    "    print(\"Reactions in peroxisome: \" + str(per))\n",
    "    print(\"--------------------------------\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
